funnel/models/blogpost.py

```diff
@@ -0,0 +1,227 @@
+
+    voteset_id = db.Column(None, db.ForeignKey('voteset.id'), nullable=False)
+    voteset = db.relationship(Voteset, uselist=False)
+
+    commentset_id = db.Column(None, db.ForeignKey('commentset.id'), nullable=False)
+    commentset = db.relationship(
+        Commentset,
+        uselist=False,
+        lazy='joined',
+        cascade='all',
+        single_parent=True,
+        backref=db.backref('blogpost', uselist=False),
+    )
+
+    search_vector = db.deferred(
+        db.Column(
+            TSVectorType(
+                'name',
+                'title',
+                'body_text',
+                weights={'name': 'A', 'title': 'A', 'body_text': 'B'},
+                regconfig='english',
+                hltext=lambda: db.func.concat_ws(
+                    ' / ', Blogpost.title, Blogpost.body_html,
+                ),
+            ),
+            nullable=False,
+        )
+    )
+
+    __table_args__ = (
+        # FIXME: Should we check for user_id as well? That would allow users to
+        # post blog posts in future in their own profile
+        db.CheckConstraint(
+            db.case([(profile_id.isnot(None), 1)], else_=0)
+            + db.case([(project_id.isnot(None), 1)], else_=0)
+            == 1,
+            name='blogpost_owner_check',
+        ),
+    )
+
+    __roles__ = {
+        'all': {
+            'read': {
+                'body',
+                'created_at',
+                'edited_at',
+                'name',
+                'title',
+                'user',
+                'visibility',
+            }
+        }
+    }
+
+    __datasets__ = {
+        'primary': {
+            'body',
+            'created_at',
+            'edited_at',
+            'name',
+            'title',
+            'user',
+            'visibility',
+        }
+    }
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+        self.voteset = Voteset(settype=SET_TYPE.BLOGPOST)
+        self.commentset = Commentset(settype=SET_TYPE.BLOGPOST)
+
+    def __repr__(self):
+        return '<Blogpost "{title}" {uuid_b58}'.format(
+            title=self.title, uuid_b58=self.uuid_b58
+        )
+
+    @with_roles(call={'author', 'profile_admin', 'project_editor'})
+    @state.transition(
+        state.DRAFT,
+        state.PUBLISHED,
+        title=__("Publish blogpost"),
+        message=__("Blogpost has been published"),
+    )
+    def publish(self, actor):
+        if self.published_at is None:
+            self.published_by = actor
+            self.published_at = db.func.utcnow()
+        else:
+            self.edited_at = db.func.utcnow()
+
+    @with_roles(call={'author', 'profile_admin', 'project_editor'})
+    @state.transition(
+        state.PUBLISHED,
+        state.DRAFT,
+        title=__("Undo publish"),
+        message=__("Blogpost is now a draft"),
+    )
+    def undo_publish(self):
+        self.published_by = None
+        self.published_at = None
```

Comment:
Don't unset. It's permanently set.

Reply:
Add a conditional state called `UNPUBLISHED` that is `DRAFT` with `published_at != None`